frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -956,6 +956,8 @@ const IR::Node* TypeInference::preorder(IR::Declaration_Instance* decl) {
             decl, decl->arguments, simpleType->to<IR::IContainer>());
         auto type = typeAndArgs.first;
         auto args = typeAndArgs.second;
+        auto *learn = clone();
+        (void)type->apply(*learn);
```

Comment:
Can `type == nullptr;` here?  You check for it just below, which is too late.  If the *learn transform does something, do you really want to throw it away?  That seems wrong.  Maybe just want `visit(type, "type")`?

Reply:
I will push this lower, after the null check.
This pattern is used all over the place in the typechecker; the recursive call updates the typeMap as a side-effect, so the result is not used. This visit may revisit some graph nodes that were already visited, that's the reason for the clone call. I believe that visit will be upset if you try to do that.